privacy-protections/fingerprinting/canvas.js

```diff
@@ -0,0 +1,194 @@
+function createCanvas (width, height) {
+    const canvasElement = document.createElement('canvas');
+    canvasElement.width = width;
+    canvasElement.height = height;
+    canvasElement.style.display = 'none';
+    return canvasElement;
+}
+
+// Writes code input into the pixel data on a canvas, returned is a data URL of the image data for the canvas.
+function generateDataURLWithCode (codeInput, alpha = 255) {
+    const encodedCode = btoa(codeInput);
+    const canvasElement = createCanvas(16, 16);
+    const canvasContext = canvasElement.getContext('2d');
+    const imageData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+    let index = 0;
+    for (let i = 0; i < imageData.data.length; i += 4) {
+        imageData.data[i] = encodedCode.charCodeAt(index) || 255;
+        // The browser optimises away transparent pixels so set alpha to fully opaque
+        // If there's alpha it'll not be consistent even in non fp protected browsers too
+        imageData.data[i + 3] = alpha;
+        ++index;
+    }
+    canvasContext.putImageData(imageData, 0, 0);
+    return canvasElement.toDataURL();
+}
+
+const tests = [
+    // Render a random 4k canvas and then get the performance of getting the pixels
+    async () => {
+        const limit = 200;
+        const canvasElement = createCanvas(3840, 2160);
+        const canvasContext = canvasElement.getContext('2d');
+        const before = performance.now();
+        const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after = performance.now();
+        const time = after - before;
+        ok(time < limit, `Getting image data must be under ${limit}ms (${time})`);
```

Comment:
Let's add a bit more context to each test
```suggestion
        ok(time < limit, `Canvas performance: getting empty canvas image data must be under ${limit}ms (${time})`);
```